src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -246,84 +308,182 @@ impl PatternFolder<'tcx> for LiteralExpander<'tcx> {
+            _ => pat.super_fold_with(self),
         }
     }
 }
 
-impl<'tcx> Pat<'tcx> {
-    fn is_wildcard(&self) -> bool {
-        match *self.kind {
-            PatKind::Binding { subpattern: None, .. } | PatKind::Wild =>
-                true,
-            _ => false
-        }
+/// A row of a matrix. Rows of len 1 are very common, which is why `SmallVec[_; 2]`
+/// works well.
+#[derive(Debug, Clone)]
+pub struct PatStack<'p, 'tcx>(SmallVec<[&'p Pat<'tcx>; 2]>);
+
+impl<'p, 'tcx> PatStack<'p, 'tcx> {
+    pub fn from_pattern(pat: &'p Pat<'tcx>) -> Self {
+        PatStack(smallvec![pat])
+    }
+
+    fn empty() -> Self {
+        PatStack(smallvec![])
+    }
+
+    fn from_vec(vec: SmallVec<[&'p Pat<'tcx>; 2]>) -> Self {
+        PatStack(vec)
+    }
+
+    fn from_slice(s: &[&'p Pat<'tcx>]) -> Self {
+        PatStack(SmallVec::from_slice(s))
+    }
+
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    fn head<'a, 'p2>(&'a self) -> &'p2 Pat<'tcx>
+    where
+        'p: 'p2,
+    {
+        self.0[0]
+    }
+    fn iter(&self) -> impl Iterator<Item = &Pat<'tcx>> {
+        self.0.iter().map(|p| *p)
+    }
+
+    /// This computes `S(constructor, self)`. See top of the file for explanations.
+    fn specialize<'a, 'p2>(
+        &self,
+        cx: &MatchCheckCtxt<'a, 'tcx>,
+        constructor: &Constructor<'tcx>,
+        ctor_wild_subpatterns: &[&'p2 Pat<'tcx>],
+    ) -> SmallVec<[PatStack<'p2, 'tcx>; 1]>
+    where
+        'a: 'p2,
+        'p: 'p2,
+    {
+        let new_heads = specialize_one_pattern(cx, self.head(), constructor, ctor_wild_subpatterns);
+        let result = new_heads
+            .into_iter()
+            .map(|mut new_head| {
+                new_head.0.extend_from_slice(&self.0[1..]);
+                new_head
+            })
+            .collect();
+        debug!("specialize({:#?}, {:#?}) = {:#?}", self, constructor, result);
+        result
+    }
+}
+
+impl<'p, 'tcx> Default for PatStack<'p, 'tcx> {
+    fn default() -> Self {
+        PatStack(smallvec![])
+    }
+}
+
+impl<'p, 'tcx> FromIterator<&'p Pat<'tcx>> for PatStack<'p, 'tcx> {
+    fn from_iter<T>(iter: T) -> Self
+    where
+        T: IntoIterator<Item = &'p Pat<'tcx>>,
+    {
+        PatStack(iter.into_iter().collect())
     }
 }
 
-/// A 2D matrix. Nx1 matrices are very common, which is why `SmallVec[_; 2]`
-/// works well for each row.
-pub struct Matrix<'p, 'tcx>(Vec<SmallVec<[&'p Pat<'tcx>; 2]>>);
+/// A 2D matrix.
+pub struct Matrix<'p, 'tcx>(Vec<PatStack<'p, 'tcx>>);
 
 impl<'p, 'tcx> Matrix<'p, 'tcx> {
     pub fn empty() -> Self {
         Matrix(vec![])
     }
 
-    pub fn push(&mut self, row: SmallVec<[&'p Pat<'tcx>; 2]>) {
+    pub fn push(&mut self, row: PatStack<'p, 'tcx>) {
         self.0.push(row)
     }
+
+    /// Iterate over the first component of each row
+    // Can't return impl Iterator because of hidden lifetime capture.
```

Comment:
1. why can't it be `-> impl Iterator<Item=&'a Pat<'tcx>> + Captures<'p>`?
2. why do you have `'p2`?

Reply:
1. I had never heard of `Captures<'p>`, this is great !
2. I think I needed it in the past because `specialize` needs to be able to extend the `'p` lifetime and somehow this forced me to have `head` relax the lifetime as well. Maybe I got confused at the time, and you're right that it's definitely not needed anymore